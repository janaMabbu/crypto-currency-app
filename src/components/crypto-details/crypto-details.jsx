import React, { PureComponent, Fragment } from 'react'
import { connect } from 'react-redux'
import ImmutablePropTypes from 'react-immutable-proptypes'
import PropTypes from 'prop-types'
import { getCurrencyDataById, getCurrencyStatusById, loadCoin } from 'ducks/get-crypto-currencies'
import { formatCurrency, formatNumber, convertPriceToNumber } from 'ducks/helpers'

import './crypto-details.less'

export class CryptoDetails extends PureComponent {
  static propTypes = {
    currencyData: ImmutablePropTypes.map,
    currencyStatus: PropTypes.bool,
    loadCoin: PropTypes.func.isRequired
  }

  componentDidMount () {
    const { loadCoin, router } = this.props
    if(router.params && router.params.currencyId) {
      loadCoin(router.params.currencyId)
    }
    // scrolling to top on browser navigation
    window.scrollTo(0, 0)
  
  }




  render () {
    return (
        <div className="row crypto-details">
        { this.renderContent() }
        </div>
    )
  }

  renderError = () => {
     return(
      <div className="crypto-details__error">
        something went wrong loading the coin!
      </div>
    )
  }

  renderContent = () => {
    const { currencyStatus } = this.props
    if(currencyStatus === null) {
      return(
      <div className="crypto-details__loading">
       <div className="spinner-border"></div>
       <div>loading...</div>
      </div>
    )
    } else {
      return (<Fragment>
      {currencyStatus ? this.renderSuccessContent(): this.renderError()}
      </Fragment>)
    }
  }

  renderSuccessContent = () => {
    const { currencyData } = this.props
     return(
      <div className="crypto-details">
       <h4 className="heading">{currencyData.get('coin_name','')} | {currencyData.get('coin_id','')}</h4>
        <p className="crypto-details__line-item price" ><strong> Price: </strong>{ formatCurrency(currencyData.get('last_price', 0),2) }</p>
        <p className="crypto-details__line-item volume"><strong> 24-hour-volume: </strong>{ formatNumber(currencyData.get('volume_24hr',0)) }</p>
        <p className="crypto-details__line-item coin-100">$100 = { 100/convertPriceToNumber(currencyData.get('last_price', 0))} Coins</p>
        <p className="crypto-details__line-item coin-250">$250 = { 250/convertPriceToNumber(currencyData.get('last_price', 0)) } Coins</p>
        <p className="crypto-details__line-item coin-5000">$5,000 = { 5000/convertPriceToNumber(currencyData.get('last_price', 0)) } Coins</p>
      </div>
    )
  }

}
const mapStateToProps = (state, props) => {
  const currencyId = props.router.params && props.router.params.currencyId
  return {
    currencyData: getCurrencyDataById(state, currencyId),
    currencyStatus: getCurrencyStatusById(state, currencyId)
  }
}
const mapDispatchToProps = {
  loadCoin
}
export default connect(mapStateToProps, mapDispatchToProps)(CryptoDetails)
