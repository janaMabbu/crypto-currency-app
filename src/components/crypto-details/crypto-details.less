@import "~less/common.less";

.crypto-details {
  padding: 50px;

  @media screen and (max-width: 576px) {
    padding: 15px;
  }

  &__line-item {
    margin: 0 0 10px;
  }

  &__error {
    text-align: center;
    padding: 30px;
    color:@red;
    font-weight: 500;
    font-size: 30px;
  }

  &__loading {
    text-align: center;
    padding: 50px;
  }
}
