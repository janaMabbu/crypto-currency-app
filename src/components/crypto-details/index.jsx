export default from './crypto-details'