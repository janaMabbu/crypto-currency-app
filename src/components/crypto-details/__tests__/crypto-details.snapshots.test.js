import { shallow } from 'enzyme'
import React from 'react'
import Immutable from 'immutable'
import { CryptoDetails } from 'components/crypto-details/crypto-details'

describe('Component - CryptoDetails', () => {
  let props
  beforeEach(() => {
    window.scrollTo = jest.fn()
    props = {
      currencyData: Immutable.fromJS({
        vol_24hr_pcnt:'-27.21',
        last_price:'9705.18401230',
        success:true,
        utc_date:'2020-02-22 21:41:01',
        coin_id:'BTC',
        currency:'USD',
        volume_24hr:'4126767296',
        time_stamp:'1582407661',
        price_24hr_pcnt:'0.50',
        currency_name:'United States Dollar',
        coin_name:'Bitcoin',
        source:'BraveNewCoin'
      }),
      currencyStatus: true,
      loadCoin: jest.fn(),
      router: {
        params: {
          currencyId:'btc',
        }
      }
    }
  })
  it('should render the CryptoDetails snapshot', () => {
    
    const wrapper = shallow(<CryptoDetails {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should only display title and content2', () => {
    const wrapper = shallow(<CryptoDetails {...props}/>)
    expect(wrapper.find('.heading').length).toBe(1)
    expect(wrapper.find('.price').length).toBe(1)
    // expect(wrapper.find('.24_hour_volume').length).toBe(1)
    // expect(wrapper.find('.100_coin').length).toBe(1)
    // expect(wrapper.find('.250_coin').length).toBe(1)
    // expect(wrapper.find('.5000_coin').length).toBe(1)
  })

  it('should render the CryptoDetails fallback', () => {
    const wrapper = shallow(<CryptoDetails {...props} />)
    wrapper.setProps({ currencyStatus: false })
    expect(wrapper.contains('something went wrong loading the coin!')).toBe(true)
  })
})