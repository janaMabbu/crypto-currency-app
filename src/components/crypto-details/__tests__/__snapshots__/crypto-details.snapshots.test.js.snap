// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - CryptoDetails should render the CryptoDetails snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CryptoDetails
    currencyData={
      Immutable.Map {
        "vol_24hr_pcnt": "-27.21",
        "last_price": "9705.18401230",
        "success": true,
        "utc_date": "2020-02-22 21:41:01",
        "coin_id": "BTC",
        "currency": "USD",
        "volume_24hr": "4126767296",
        "time_stamp": "1582407661",
        "price_24hr_pcnt": "0.50",
        "currency_name": "United States Dollar",
        "coin_name": "Bitcoin",
        "source": "BraveNewCoin",
      }
    }
    currencyStatus={true}
    loadCoin={
      [MockFunction] {
        "calls": Array [
          Array [
            "btc",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    router={
      Object {
        "params": Object {
          "currencyId": "btc",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <React.Fragment>
        <div
          className="crypto-details"
        >
          <h4
            className="heading"
          >
            Bitcoin
             | 
            BTC
          </h4>
          <p
            className="crypto-details__line-item price"
          >
            <strong>
               Price: 
            </strong>
            $9,705.18
          </p>
          <p
            className="crypto-details__line-item 24_hour_volume"
          >
            <strong>
               24-hour-volume: 
            </strong>
            4,126,767,296
          </p>
          <p
            className="crypto-details__line-item 100_coin"
          >
            $100 = 
            0.010303771661955468
             Coins
          </p>
          <p
            className="crypto-details__line-item 250_coin"
          >
            $250 = 
            0.025759429154888668
             Coins
          </p>
          <p
            className="crypto-details__line-item 5000_coin"
          >
            $5,000 = 
            0.5151885830977734
             Coins
          </p>
        </div>
      </React.Fragment>,
      "className": "row crypto-app",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="crypto-details"
        >
          <h4
            className="heading"
          >
            Bitcoin
             | 
            BTC
          </h4>
          <p
            className="crypto-details__line-item price"
          >
            <strong>
               Price: 
            </strong>
            $9,705.18
          </p>
          <p
            className="crypto-details__line-item 24_hour_volume"
          >
            <strong>
               24-hour-volume: 
            </strong>
            4,126,767,296
          </p>
          <p
            className="crypto-details__line-item 100_coin"
          >
            $100 = 
            0.010303771661955468
             Coins
          </p>
          <p
            className="crypto-details__line-item 250_coin"
          >
            $250 = 
            0.025759429154888668
             Coins
          </p>
          <p
            className="crypto-details__line-item 5000_coin"
          >
            $5,000 = 
            0.5151885830977734
             Coins
          </p>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h4
              className="heading"
            >
              Bitcoin
               | 
              BTC
            </h4>,
            <p
              className="crypto-details__line-item price"
            >
              <strong>
                 Price: 
              </strong>
              $9,705.18
            </p>,
            <p
              className="crypto-details__line-item 24_hour_volume"
            >
              <strong>
                 24-hour-volume: 
              </strong>
              4,126,767,296
            </p>,
            <p
              className="crypto-details__line-item 100_coin"
            >
              $100 = 
              0.010303771661955468
               Coins
            </p>,
            <p
              className="crypto-details__line-item 250_coin"
            >
              $250 = 
              0.025759429154888668
               Coins
            </p>,
            <p
              className="crypto-details__line-item 5000_coin"
            >
              $5,000 = 
              0.5151885830977734
               Coins
            </p>,
          ],
          "className": "crypto-details",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Bitcoin",
                " | ",
                "BTC",
              ],
              "className": "heading",
            },
            "ref": null,
            "rendered": Array [
              "Bitcoin",
              " | ",
              "BTC",
            ],
            "type": "h4",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <strong>
                   Price: 
                </strong>,
                "$9,705.18",
              ],
              "className": "crypto-details__line-item price",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": " Price: ",
                },
                "ref": null,
                "rendered": " Price: ",
                "type": "strong",
              },
              "$9,705.18",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <strong>
                   24-hour-volume: 
                </strong>,
                "4,126,767,296",
              ],
              "className": "crypto-details__line-item 24_hour_volume",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": " 24-hour-volume: ",
                },
                "ref": null,
                "rendered": " 24-hour-volume: ",
                "type": "strong",
              },
              "4,126,767,296",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "$100 = ",
                0.010303771661955468,
                " Coins",
              ],
              "className": "crypto-details__line-item 100_coin",
            },
            "ref": null,
            "rendered": Array [
              "$100 = ",
              0.010303771661955468,
              " Coins",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "$250 = ",
                0.025759429154888668,
                " Coins",
              ],
              "className": "crypto-details__line-item 250_coin",
            },
            "ref": null,
            "rendered": Array [
              "$250 = ",
              0.025759429154888668,
              " Coins",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "$5,000 = ",
                0.5151885830977734,
                " Coins",
              ],
              "className": "crypto-details__line-item 5000_coin",
            },
            "ref": null,
            "rendered": Array [
              "$5,000 = ",
              0.5151885830977734,
              " Coins",
            ],
            "type": "p",
          },
        ],
        "type": "div",
      },
      "type": Symbol(react.fragment),
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <React.Fragment>
          <div
            className="crypto-details"
          >
            <h4
              className="heading"
            >
              Bitcoin
               | 
              BTC
            </h4>
            <p
              className="crypto-details__line-item price"
            >
              <strong>
                 Price: 
              </strong>
              $9,705.18
            </p>
            <p
              className="crypto-details__line-item 24_hour_volume"
            >
              <strong>
                 24-hour-volume: 
              </strong>
              4,126,767,296
            </p>
            <p
              className="crypto-details__line-item 100_coin"
            >
              $100 = 
              0.010303771661955468
               Coins
            </p>
            <p
              className="crypto-details__line-item 250_coin"
            >
              $250 = 
              0.025759429154888668
               Coins
            </p>
            <p
              className="crypto-details__line-item 5000_coin"
            >
              $5,000 = 
              0.5151885830977734
               Coins
            </p>
          </div>
        </React.Fragment>,
        "className": "row crypto-app",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <div
            className="crypto-details"
          >
            <h4
              className="heading"
            >
              Bitcoin
               | 
              BTC
            </h4>
            <p
              className="crypto-details__line-item price"
            >
              <strong>
                 Price: 
              </strong>
              $9,705.18
            </p>
            <p
              className="crypto-details__line-item 24_hour_volume"
            >
              <strong>
                 24-hour-volume: 
              </strong>
              4,126,767,296
            </p>
            <p
              className="crypto-details__line-item 100_coin"
            >
              $100 = 
              0.010303771661955468
               Coins
            </p>
            <p
              className="crypto-details__line-item 250_coin"
            >
              $250 = 
              0.025759429154888668
               Coins
            </p>
            <p
              className="crypto-details__line-item 5000_coin"
            >
              $5,000 = 
              0.5151885830977734
               Coins
            </p>
          </div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h4
                className="heading"
              >
                Bitcoin
                 | 
                BTC
              </h4>,
              <p
                className="crypto-details__line-item price"
              >
                <strong>
                   Price: 
                </strong>
                $9,705.18
              </p>,
              <p
                className="crypto-details__line-item 24_hour_volume"
              >
                <strong>
                   24-hour-volume: 
                </strong>
                4,126,767,296
              </p>,
              <p
                className="crypto-details__line-item 100_coin"
              >
                $100 = 
                0.010303771661955468
                 Coins
              </p>,
              <p
                className="crypto-details__line-item 250_coin"
              >
                $250 = 
                0.025759429154888668
                 Coins
              </p>,
              <p
                className="crypto-details__line-item 5000_coin"
              >
                $5,000 = 
                0.5151885830977734
                 Coins
              </p>,
            ],
            "className": "crypto-details",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Bitcoin",
                  " | ",
                  "BTC",
                ],
                "className": "heading",
              },
              "ref": null,
              "rendered": Array [
                "Bitcoin",
                " | ",
                "BTC",
              ],
              "type": "h4",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <strong>
                     Price: 
                  </strong>,
                  "$9,705.18",
                ],
                "className": "crypto-details__line-item price",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": " Price: ",
                  },
                  "ref": null,
                  "rendered": " Price: ",
                  "type": "strong",
                },
                "$9,705.18",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <strong>
                     24-hour-volume: 
                  </strong>,
                  "4,126,767,296",
                ],
                "className": "crypto-details__line-item 24_hour_volume",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": " 24-hour-volume: ",
                  },
                  "ref": null,
                  "rendered": " 24-hour-volume: ",
                  "type": "strong",
                },
                "4,126,767,296",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "$100 = ",
                  0.010303771661955468,
                  " Coins",
                ],
                "className": "crypto-details__line-item 100_coin",
              },
              "ref": null,
              "rendered": Array [
                "$100 = ",
                0.010303771661955468,
                " Coins",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "$250 = ",
                  0.025759429154888668,
                  " Coins",
                ],
                "className": "crypto-details__line-item 250_coin",
              },
              "ref": null,
              "rendered": Array [
                "$250 = ",
                0.025759429154888668,
                " Coins",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "$5,000 = ",
                  0.5151885830977734,
                  " Coins",
                ],
                "className": "crypto-details__line-item 5000_coin",
              },
              "ref": null,
              "rendered": Array [
                "$5,000 = ",
                0.5151885830977734,
                " Coins",
              ],
              "type": "p",
            },
          ],
          "type": "div",
        },
        "type": Symbol(react.fragment),
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
